services/src/main/java/org/keycloak/protocol/ciba/utils/DecoupledAuthnResultParser.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.protocol.ciba.utils;
+
+import java.util.Map;
+import java.util.UUID;
+
+import org.jboss.logging.Logger;
+import org.keycloak.common.util.Time;
+import org.keycloak.models.CodeToTokenStoreProvider;
+import org.keycloak.models.KeycloakSession;
+
+public class DecoupledAuthnResultParser {
+
+    private static final Logger logger = Logger.getLogger(DecoupledAuthnResultParser.class);
+
+    public static void persistDecoupledAuthnResult(KeycloakSession session, String id, DecoupledAuthnResult decoupledAuthnResultData, int expires_in) {
+        CodeToTokenStoreProvider codeStore = session.getProvider(CodeToTokenStoreProvider.class);
+
+        if (id == null) {
+            throw new IllegalStateException("ID not present in the data");
+        }
+        UUID key = UUID.fromString(id);
+
+        Map<String, String> serialized = decoupledAuthnResultData.serializeCode();
+        codeStore.put(key, expires_in, serialized);
```

Comment:
I think `expires_in` is in milliseconds and the provider is going to set expiration based on seconds. That would cause entries to take a lot of time to expire, and not really what you are expected.

See https://github.com/keycloak/keycloak/blob/80f4bd382225d35b6177ce2c9b7ab7fa84e0d1fd/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanCodeToTokenStoreProvider.java#L54.

Reply:
@pedroigor I've supposed that this `expires_in` is in seconds. And I've written codes for treating it in seconds. I could not found out the point I've treated this `expires_in` in milliseconds.

Reply:
Sorry, maybe I read the code too quickly then.